[minifier] Fix config generator for callables

yushangdi opened the pull request:
Summary:
When config contains callables, the current configs generated cannot be run:

```
torch._dynamo.config.reorderable_logging_functions = {<built-in function print>, <function warning at 0x7f774c595630>, <function log at 0x7f774c595870>, <function error at 0x7f774c595510>, <function info at 0x7f774c595750>, <built-in function warn>, <function exception at 0x7f774c5955a0>, <function debug at 0x7f774c5957e0>, <function critical at 0x7f774c5953f0>}
```

We fix the config to generate the right string, so the config is runnable, like below

```
import logging
import warnings
torch._dynamo.config.reorderable_logging_functions = { warnings.warn, logging.warn, print }
```

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:utils -- -r test_codegen_config
```

Differential Revision: D67998703

desertfire commented on torch/utils/_config_module.py:
```diff
@@ -493,12 +493,61 @@ def codegen_config(self) -> str:
+        imports = set()
+
+        def get_module_name(func, add_dot):
+            module_name = func.__module__
```
Not sure why simply returning module_name is not enough here.

yushangdi replied:
> Not sure why simply returning module_name is not enough here.

@desertfire For python native functions like `print`, the module name is `builtins`, but `builtins.print` is not a runnable command. So we need to remove the `builtins` part.

For `_warnings`, when you import the `warnings` module, Python actually imports a module called `_warnings` internally. The `warnings` module is a wrapper around the `_warnings` module that provides a more user-friendly interface.

In general, one should use the warnings module, rather than trying to access the _warnings module directly.

desertfire replied:
I suspect `_warnings` will use the same API. Especially as debugging tool, we don't have to worry about API BC/FC here.

Skylion007 commented on torch/testing/_internal/fake_config_module.py:
```diff
@@ -28,6 +28,7 @@
 e_aliased_bool: bool = Config(
     alias="torch.testing._internal.fake_config_module2.e_aliasing_bool"
 )
+e_func: Callable = None
```
Surprised ruff ruf013 didn't immediately flag and correct this.

yushangdi replied:
fixed now!

desertfire commented on torch/utils/_config_module.py:
```diff
@@ -493,12 +493,61 @@ def codegen_config(self) -> str:
+                module_name = ""
+            if func.__module__ == "_warnings":
+                module_name = "warnings"
+            if add_dot and module_name != "":
```
How about this one?  Can we turn relative import into absolute import?

yushangdi replied:
> How about this one? Can we turn relative import into absolute import?

@desertfire  The "dot" here is the dot between "loggings.warn". It's not for the absolute/relative import. I have this flag here because if the module is builtin, then we should just have "print", but not ".print".

desertfire replied:
I see. Then I think spelling out the full `builtins.print` makes more sense here. As a debugging tool, it's fine that the generated file is a little verbose.

yushangdi replied:
> I see. Then I think spelling out the full `builtins.print` makes more sense here. As a debugging tool, it's fine that the generated file is a little verbose.

@desertfire We cannot spell out the full `builtins.print`, because `builtins.print` is not runnable. You'll get `NameError: name 'builtins' is not defined` error. 

btw I made some minor changes to the PR to check `isinstance(v, __module__)`. If `v` is a partial function, for example, `functools.partial(<function collect_fwd_graph_outputs at 0x7fbfe640e050>, fwd_outputs=set())`, it won't have a module. 

In this case, we'll just output the string above. It's not runnable, I can fix that later in a followup PR.

yushangdi committed:
[minifier] Fix config generator for callables (#144518)

Summary:

When config contains callables, the current configs generated cannot be run:

```
torch._dynamo.config.reorderable_logging_functions = {<built-in function print>, <function warning at 0x7f774c595630>, <function log at 0x7f774c595870>, <function error at 0x7f774c595510>, <function info at 0x7f774c595750>, <built-in function warn>, <function exception at 0x7f774c5955a0>, <function debug at 0x7f774c5957e0>, <function critical at 0x7f774c5953f0>}
```

We fix the config to generate the right string, so the config is runnable, like below

```
import logging
import warnings
torch._dynamo.config.reorderable_logging_functions = { warnings.warn, logging.warn, print }
```

Test Plan:
```
buck2 run 'fbcode//mode/dev-nosan' fbcode//caffe2/test:utils  -- -r TestConfigModule

python test/dynamo/test_activation_checkpointing.py ActivationCheckpointingViaTagsTests.test_tags_must_save_tensor_that_has_backward_hook
```

Reviewed By: desertfire

Differential Revision: D67998703

izaitsevfb commented:
@pytorchbot merge

yushangdi commented:
@pytorchbot merge -i
